### PR TITLE
Clean up some of the config logic

### PR DESCRIPTION
**Pull request by bryce-anderson**
Break down the extraction process.
Assert structure better.
Avoid panicing on bad config.

### CODE REVIEW

**Comment by bryce-anderson** [comment]
This is not quite complete: I want to put some tests in there (mostly to make sure I didn't break anything). Feedback welcome.

**Comment by brayniac** [comment]
I will review this tomorrow. +1 on adding tests.

**Comment by bryce-anderson** [comment]
I've added a test case which parses the `thrift_calc.toml` example and checks for consistency.

**Comment by bryce-anderson** [comment]
As a note, my motivation for this is to facilitate #64: by making parameter extraction a function it is possible to recurse making the extraction of a recursive parameter model relatively trivial.

File: src/config.rs [hunk 1]
```diff
@@ -68,160 +72,206 @@ impl Default for BenchmarkConfig {
...
+    assert_eq!(workload.workloads.len(), 3);
+
+    let w0 = &workload.workloads[0];
+    // Check the first workload
+    assert_eq!(w0.method, "ping");
+    assert_eq!(w0.rate, 1);
+    assert_eq!(w0.parameters.len(), 0);
+
+    let w2 = &workload.workloads[2];
+    // check the second workload
```
Comment (brayniac) [inline src/config.rs]: nit: w2 is the third workload

File: src/config.rs [hunk 2]
```diff
@@ -68,160 +72,206 @@ impl Default for BenchmarkConfig {
...
+    assert_eq!(w2.method, "calculate");
+    assert_eq!(w2.rate, 1);
+    assert_eq!(w2.parameters.len(), 6);
+
+    // Check that the first parameter of the second workload was parsed correctly
```
Comment (brayniac) [inline src/config.rs]: same as line 261

File: src/config.rs [hunk 3]
```diff
@@ -68,160 +72,206 @@ impl Default for BenchmarkConfig {
...
+        Some("struct") => Type::Struct,
+        Some("map") => Type::Map,
+        Some("set") => Type::Set,
+        Some("list") => {
+            Type::List(parameter.get("contains")
+                                .and_then(|k| k.as_str())
+                                .unwrap()
+                                .to_owned())
+        }
+        Some(_) => Type::None,
```
Comment (brayniac) [inline src/config.rs]: I wonder if we should ```Some(other) => return Err(...)``` for this case. Matches existing behavior, so up to you.

File: src/config.rs (hunk 1 above)
Comment (bryce-anderson) [inline src/config.rs]: Can do: I left the numbers in 0 indexed form but I'm happy to change them to counting numbers.

File: src/config.rs (hunk 3 above)
Comment (bryce-anderson) [inline src/config.rs]: I'm happy to change the behavior to result in a failure: I think an error is more correct in this situation.

File: src/config.rs (hunk 1 above)
Comment (brayniac) [inline src/config.rs]: Zero indexed is fine. It's just that this config is ping, add, subtract - if ping (w0) is first, subtract (w2) is third (not second). The comment led me to think you were talking about the second workload in the config, not the third.

Comment (bryce-anderson) [inline src/config.rs]: Woops... I misinterpreted your comment: it was for the comment, not the name.

**Comment by brayniac** [comment]
First off, thanks for deciding to tackle this. Definitely needed to fix #64 

Overall, this looks great. Happy to merge once the comments are addressed and you're happy with it.

**Comment by bryce-anderson** [comment]
I've addressed your concerns and changed the name of `extract_parameters` to be singular.

I'm happy with it once Travis clears it.

**Pull request merged**